### PR TITLE
Simple crossover strategy

### PR DESCRIPTION
**Pull request by james-bc**

### CODE REVIEW

File: bcex/examples/simple_strategy.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+import pandas as pd
+import pytz
+from bcex.core.orders import OrderSide, OrderType
+from bcex.core.websocket_client import Environment
+from bcex.examples.trader import BaseTrader
+from requests import get
+
+
+def datetime2unixepoch(dt):
```
Comment (simon-bc) [inline bcex/examples/simple_strategy.py:12]: not here

File: bcex/examples/simple_strategy.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    dt: datetime
+
+    Returns
+    -------
+    int
+        unix epoch
+    """
+    if dt.tzinfo is None:
+        dt = pytz.UTC.localize(dt)
+    else:
```
Comment (simon-bc) [inline bcex/examples/simple_strategy.py:26]: should we not use datetime.timestamp() ?

File: bcex/examples/simple_strategy.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        self.rolling_window = rolling_window
+        self.granularity = granularity
+        self._historical_candles = None
+        self.start_date = start_date
+        self.balance_percentage = balance_percentage
+        self.latest_timestamp = None
+
+    def get_historical_candles(self):
+        end_date = datetime.now(pytz.UTC)
+        prices_url = (
```
Comment (simon-bc) [inline bcex/examples/simple_strategy.py:87]: use requests with params

File: bcex/examples/simple_strategy.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            refresh_rate=refresh_rate,
+            env=env,
+            channels_kwargs=channel_kwargs,
+            **kwargs,
+        )
+        self.rolling_window = rolling_window
+        self.granularity = granularity
+        self._historical_candles = None
+        self.start_date = start_date
+        self.balance_percentage = balance_percentage
```
Comment (simon-bc) [inline bcex/examples/simple_strategy.py:82]: you are using it as a fraction?

Comment (james-bc) [reply bcex/examples/simple_strategy.py:82]: ye to place a % of the relevant balance? can change it if we want

Comment (tduriez-bc) [reply bcex/examples/simple_strategy.py:37]: I think he means you should rename into `balance_fraction`

File: bcex/core/exchange_interface.py [hunk 5]
```diff
@@ -487,6 +492,19 @@ def get_available_balance(self, coin):
...
+
+        Parameters
+        ----------
+        symbol : Symbol
+
+        Returns
+        -------
+        order_book : Dict
+        """
+        return self.ws.l2_book[symbol]
```
Comment (tduriez-bc) [inline bcex/core/exchange_interface.py:506]: you'll have a conflict here, I renamed the attribute `ws` into `client`

File: bcex/core/exchange_interface.py [hunk 6]
```diff
@@ -16,7 +16,12 @@ class ExchangeInterface:
         websocket client to handle interactions with the exchange
     """
 
-    REQUIRED_CHANNELS = [Channel.SYMBOLS, Channel.TICKER, Channel.TRADES]
+    REQUIRED_CHANNELS = [
+        Channel.SYMBOLS,
+        Channel.TICKER,
+        Channel.TRADES,
+        Channel.HEARTBEAT,
```
Comment (tduriez-bc) [inline bcex/core/exchange_interface.py:23]: FYI I've fixed the error about heartbeat, so you don't need to include

File: bcex/core/utils.py [hunk 7]
```diff
@@ -22,3 +25,46 @@ def update_max_list(l, e, n):
     """Update a list and enforce maximum length"""
     l.append(e)
     return l[-n:]
+
+
+def datetime2unixepoch(dt):
+    """ Utility to transform a datetime instance into an unix epoch represented as an int.
```
Comment (simon-bc) [inline bcex/core/utils.py:31]: this is milliseconds - you should make that clear

File: bcex/core/utils.py [hunk 8]
```diff
@@ -22,3 +25,46 @@ def update_max_list(l, e, n):
...
+    return int(result)
+
+
+def unixepoch2datetime(unixepoch):
+    """Given a timestamp, it returns a datetime object.
+
+    Parameters
+    ----------
+    unixepoch : int
+       Timestamp. It can have either a second or millisecond precision
```
Comment (simon-bc) [inline bcex/core/utils.py:59]: it cannot have second

File: bcex/core/websocket_client.py [hunk 9]
```diff
@@ -609,17 +609,17 @@ def _on_l2_updates(self, msg):
...
                         self.l2_book[symbol][book][price] = size
         else:
             self._on_unsupported_event_message(msg, Channel.L2)
 
         if len(self.l2_book[symbol][Book.ASK]) > 0:
-            logging.info(f"Ask: {self.l2_book[symbol][Book.ASK].peekitem(0)} ")
+            logging.debug(f"Ask: {self.l2_book[symbol][Book.ASK].peekitem(0)} ")
         if len(self.l2_book[symbol][Book.BID]) > 0:
-            logging.info(f"Bid: {self.l2_book[symbol][Book.BID].peekitem(-1)}")
+            logging.debug(f"Bid: {self.l2_book[symbol][Book.BID].peekitem(-1)}")
```
Comment (simon-bc) [inline bcex/core/websocket_client.py:622]: good stuff - we should clean up the logging 👍

File: bcex/examples/simple_strategy.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            ).T
+            return df_res
+        return pd.DataFrame()
+
+    @property
+    def live_candles(self):
+        return self.get_latest_candles()
+
+    def get_candle_df(self):
+        df = pd.concat([self.historical_candles, self.live_candles]).sort_index()
```
Comment (simon-bc) [inline bcex/examples/simple_strategy.py:92]: potential for duplicate indexes?

File: bcex/core/websocket_client.py [hunk 11]
```diff
@@ -295,6 +295,7 @@ def _send_subscriptions_to_ws(self, channels):
             else:
                 subscription = {"action": Action.SUBSCRIBE, "channel": channel}
                 if kwargs:
+                    print(kwargs)
```
Comment (tduriez-bc) [inline bcex/core/websocket_client.py:298]: remove

**Comment by simon-bc** [comment]
> Code lgtm.
> 
> Maybe will need refining on the candles.
> 
> We've seen that there could be duplicated timestamps in candles received, suggesting the current candle could be updated live until it's done ?
> It means you could store and make decision from unfinished candle(s) 📉
> 
> Example of duplicates candles :
> 
> ```
> 'price': [1588629180000, 8939.9, 8940.4, 8939.9, 8940.4, 0.0056]}
> 'price': [1588629180000, 8939.9, 8940.4, 8939.9, 8939.9, 0.42581387]
> ```

you could always add a check to see if the candle is complete i.e. you have the candles start time, check time now, and if difference is greater than granularity, you are good to go

**Comment by james-bc** [comment]
ye we probably need to add something to the prices channel in the client to update the candle instead of print a new one, probably shouldnt be on the strategy

**Pull request merged**